### PR TITLE
If applied this commit denotes document boundaries on Dispersion Plot viz

### PR DESCRIPTION
**Pull request by lwgray**
This is a partial fix to issue #549 for which I am attempting to define document boundaries within a dispersion plot.

I had to make some minor changes to make this work.  The visualizer was changed so that the expected input is now a list of documents.  Before the input was a single list representing all the words of the corpora.  This change makes it easier to define the beginning/end of a document and keep track of the word offset.

PLOTS
---------
This plot is of the occurrence of these words in documents from the 'sporting' subset of the hobbies dataset.  The light grey lines denote the document boundaries.

image

Coding (greatly appreciated)
-----
* my variable naming can be improved, any suggestions??? 

Visuals
-------
* Suggestions on alpha, linecolor, linestyle, and linewidth

### CODE REVIEW

File: yellowbrick/text/dispersion.py [hunk 1]
```diff
@@ -47,20 +47,24 @@ class DispersionPlot(TextVisualizer):
...
-    def __init__(self, words, ax=None, color=None, ignore_case=False, **kwargs):
+    def __init__(self, words, ax=None, color=None, ignore_case=False,
+                 boundary=False, **kwargs):
         super(DispersionPlot, self).__init__(ax=ax, **kwargs)
         
         self.color = color
         self.words = words
         self.ignore_case = ignore_case
-        
+        self.set_boundaries = boundary
```
Comment (bbengfort) [inline yellowbrick/text/dispersion.py]: I think it's important that the properties on the visualizer match the variables that are passed in the arg list; Scikit-Learn using metaprogramming to detect what's in the `__init__` args for `get_params` and `set_params`, which we also make use of.

File: yellowbrick/text/dispersion.py [hunk 2]
```diff
@@ -47,20 +47,24 @@ class DispersionPlot(TextVisualizer):
     ignore_case : boolean, default: False
 	Specify whether input  will be case-sensitive.
         
+    boundary : boolean, default: False
```
Comment (bbengfort) [inline yellowbrick/text/dispersion.py]: I guess I don't mind `boundaries` or even just `annotate_docs`? It might be important to specify whether the end or the beginning of the document is being annotated in the comment.

File: yellowbrick/text/dispersion.py [hunk 3]
```diff
@@ -84,13 +88,25 @@ def fit(self, text):
             A list of words in the order they appear in the corpus.
         """
             
+        # Concatenating documents and define boundaries by word offset
+        self.boundaries_ = []
+        self.doc_words_ = []
+        self.offset = 0
+
+        for doc in text:
+            for word in doc.split():
```
Comment (bbengfort) [inline yellowbrick/text/dispersion.py]: I think it's reasonable to expect as input a list of lists; e.g. a list of documents where each document is a list of words as opposed to a list of strings. This is closer to what scikit-learn does in `sklearn.feature_extraction.text`, which we'll want to use eventually in this module, e.g. #488 and will also allow the user to use `nltk.word_tokenize` or `nltk.bigrams` if required. 

Therefore I recommend changing this to `for word in doc:` rather than simply splitting on space.

File: yellowbrick/text/dispersion.py [hunk 4]
```diff
@@ -84,13 +88,25 @@ def fit(self, text):
             A list of words in the order they appear in the corpus.
         """
             
+        # Concatenating documents and define boundaries by word offset
+        self.boundaries_ = []
+        self.doc_words_ = []
+        self.offset = 0
```
Comment (bbengfort) [inline yellowbrick/text/dispersion.py]: We don't really want to store the entire corpus in memory on the visualizer, it's possible that it's an extremely big corpus that's being stored on disk and a generator is reading through all the documents; which is why we had the generator test to ensure that we only looped through the input once (this loops through twice). 

Is it possible to compute the boundaries and the points in one single loop?

File: yellowbrick/text/dispersion.py [hunk 5]
```diff
@@ -103,6 +119,11 @@ def draw(self, points, **kwargs):
         kwargs: generic keyword arguments.
         """
         
+        # Define boundaries with a solid vertical line
+        if self.set_boundaries:
+            for xcoords in self.boundaries_:
+                self.ax.axvline(x=xcoords, color='lightgray', alpha=0.75,
```
Comment (bbengfort) [inline yellowbrick/text/dispersion.py]: I like that you're trying to mute the document boundaries so there is a contrast between the words and the annotations - very nice! Perhaps setting `linestyle='--'` or `':'` would also help?

File: yellowbrick/text/dispersion.py (hunk 2 above)
Comment (lwgray) [reply yellowbrick/text/dispersion.py]: I have decided to go with annotate_docs

File: yellowbrick/text/dispersion.py [hunk 6]
```diff
@@ -31,69 +31,80 @@ class DispersionPlot(TextVisualizer):
...
-	    # NOTE: this will find all indices if duplicate words are supplied 
-	    # In the case that word is not in target words, any empty list is
-	    # returned and no data will be yielded 
-            for y in (self.target_words_ == word).nonzero()[0]:
-                yield (x, y)
+    def _compute_dispersion(self, text):
+        self.boundaries_ = []
+        self.offset = 0
+        for doc in text:
+            for word in doc:
```
Comment (lwgray) [inline yellowbrick/text/dispersion.py]: here, I define boundaries and points in a single loop

**Comment by lwgray** [comment]
I will add a test for boundaries in a follow-up commit

File: tests/test_text/test_dispersion.py [hunk 7]
```diff
@@ -67,12 +67,27 @@ def test_dispersionplot_generator_input(self):
         """
         corpus = self.load_data('hobbies')
 
-        text = chain(*map(lambda s: s.split(), corpus.data))
+        text = (doc.split() for doc in corpus.data)
```
Comment (bbengfort) [inline tests/test_text/test_dispersion.py]: We do still need the `map` here to ensure a generator is passed in; this line of code is creating a tuple in memory. Would you please change it to:

```python
text = map(lambda doc: doc.split() for doc in corpus.data)
```

File: tests/test_text/test_dispersion.py [hunk 8]
```diff
@@ -67,12 +67,27 @@ def test_dispersionplot_generator_input(self):
...
+        corpus = self.load_data('hobbies')
+
+        text = (doc.split() for doc in corpus.data)
+        target_words = ['girl', 'she', 'boy', 'he', 'man']
+
+        visualizer = DispersionPlot(target_words, annotate_docs=True)
+        visualizer.fit(text)
+        visualizer.ax.grid(False)
+
+        self.assert_images_similar(visualizer, tol=25)
```
Comment (bbengfort) [inline tests/test_text/test_dispersion.py]: Good test, thank you for adding it!

File: tests/test_text/test_dispersion.py [hunk 9]
```diff
@@ -21,7 +21,7 @@
 from yellowbrick.text.dispersion import *
 from tests.dataset import DatasetMixin
 from tests.base import VisualTestCase
-from itertools import chain
+# from itertools import chain
```
Comment (bbengfort) [inline tests/test_text/test_dispersion.py]: You can go ahead and delete this -- we won't need it with the new input.

File: yellowbrick/text/dispersion.py [hunk 10]
```diff
@@ -31,69 +31,82 @@ class DispersionPlot(TextVisualizer):
...
         """
         The fit method is the primary drawing input for the dispersion
         visualization. It requires the corpus as a list of words.
-        
+
         Parameters
         ----------
         text : list
-            A list of words in the order they appear in the corpus.
+            Should be provided as a list of documents that contain
```
Comment (bbengfort) [inline yellowbrick/text/dispersion.py]: Thanks for updating the docstring!

File: tests/test_text/test_dispersion.py (hunk 7 above)
Comment (bbengfort) [reply tests/test_text/test_dispersion.py]: Yes, you're right - sorry; I mistook the parens for a tuple comprehension, but that does indeed wrap a generator expression. Might be a good comment, but you're right, go ahead and merge this in!

https://stackoverflow.com/questions/16940293/why-is-there-no-tuple-comprehension-in-python
